### PR TITLE
[storage.blob] Remove requests as dependency for storage.blob

### PR DESCRIPTION
**Pull request by Stevenjin8**
# Description

Similar to #24965. This removes the `requests` import from the `storage.blob` client and (hopefully) fixes the retry mechanism (see comments)

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md).**

## General Guidelines and Best Practices
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md##building-and-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/storage/azure-storage-blob/azure/storage/blob/_download.py
Comment (Stevenjin8) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_download.py:212]: I think this bit of code is trying to catch exceptions generated by `self.client.download` (few lines above this).

Comment (Stevenjin8) [reply sdk/storage/azure-storage-blob/azure/storage/blob/_download.py:212]: same idea below

File: sdk/storage/azure-storage-blob/azure/storage/blob/_download.py
Comment (Stevenjin8) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_download.py:208]: The requests transport will catch `requests.exceptions.ConnectionError` and raise it as a `ServiceResponseError` ([here](/sdk/core/azure-core/azure/core/pipeline/transport/_requests_basic.py:34))

File: sdk/storage/azure-storage-blob/azure/storage/blob/_download.py
Comment (Stevenjin8) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_download.py:214]: I think this is trying to catch exceptions raised by `self.client.download` (line 200). Same idea below.

Comment (jalauzon-msft) [reply sdk/storage/azure-storage-blob/azure/storage/blob/_download.py:214]: I had a look at this and I'm not sure moving the retry logic away from `process_content` was correct here. The call to `client.download` will actually returning an `IO` object that can stream the response from the service (if the download I big enough). This means inside `process_content` we will actually be making additional requests that I believe this retry logic is trying to catch. The requests will happen here:
https://github.com/Azure/azure-sdk-for-python/blob/b0f9e02b2d597a57dcef33bb265c4b5c47ad0a85/sdk/storage/azure-storage-blob/azure/storage/blob/_download.py#L46

I don't know the full history here, but I believe at one point the additional requests made during response streaming did not pass through our normal pipeline and therefore did not get our normal exception handling or retry policies. This would make sense why additional retry logic was added here and why raw `requests` and `aiohttp` exceptions were being caught. I'm honestly not sure if that is still the case though or if there was a change to make response streaming use the normal pipeline.

If response streaming now uses the normal pipeline, we may not even need the retries here at all and may actually be tripling our retries if we start catching the core exceptions.

@lmazuel, @xiangyan99, @iscai-msft do any of you know if response streaming uses the standard pipeline now? The generated code accepts the pipeline as a param now so that would seem to indicate it does?
https://github.com/Azure/azure-sdk-for-python/blob/b0f9e02b2d597a57dcef33bb265c4b5c47ad0a85/sdk/storage/azure-storage-blob/azure/storage/blob/_generated/operations/_blob_operations.py#L1757

Comment (Stevenjin8) [reply sdk/storage/azure-storage-blob/azure/storage/blob/_download.py:214]: I think getting rid of the retries might be the way to go. When I debug through the following code, each iteration (except for the first one) in the for loop goes into the pipeline.

```python
    account_client = BlobServiceClient(ENDPOINT, KEY)
    blob_client = BlobClient(
        account_url=account_client.url,
        container_name="tsjinxuanstorage2",
        blob_name="random",
        credential=account_client.credential,
        max_single_get_size=1,
        max_chunk_get_size=1,
    )

    downloader = blob_client.download_blob()
    for chunk in downloader.chunks():
        print(chunk)
```

Is this what you mean by an `IO` object that streams data? If so, I think you are right. The pipeline itself has a [retry policy]( https://github.com/Azure/azure-sdk-for-python/blob/1486a398c39cf9e2f1b7aa938a3e7e52f55fd6e1/sdk/storage/azure-storage-blob/azure/storage/blob/_shared/base_client.py#L243) so my change would indeed be tripling the number of retries. Also, this means that we should update #24965 accordingly.

File: sdk/storage/azure-storage-blob/CHANGELOG.md
Comment (iscai-msft) [inline sdk/storage/azure-storage-blob/CHANGELOG.md:9]: nit: we don't want to mention private modules, so it can be like 'Removed forced `requests` import for sync calls' and 'fixed blob download retry mechanism' or something like that

Comment (vincenttran-msft) [reply sdk/storage/azure-storage-blob/CHANGELOG.md:9]: Super nit: Same feedback as other PR, see line 68 (original) or line 70 (your current commit) for an example of mentioning an issue in the changelog for consistency sake 😄

File: sdk/storage/azure-storage-blob/azure/storage/blob/_download.py
Comment (jalauzon-msft) [reply sdk/storage/azure-storage-blob/azure/storage/blob/_download.py:214]: @Stevenjin8 this code will actually make a series of regular GET (`self.cleint.download`) requests that each ask for 1 Byte so response streaming would not actually kick in here. All of the data would come in the initial download call and no additional requests would happen in `process_content` AFAIK. The pipeline being triggered while debugging will come from each explicit call to download.

I'm not entirely sure what triggers response streaming, but you can see we are using it here. I guess I'd assume it would be triggered by a certain size payload from the service?
https://github.com/Azure/azure-sdk-for-python/blob/0a20bfba26ac5f43dcecb203311df63e3abe24c6/sdk/storage/azure-storage-blob/azure/storage/blob/_generated/operations/_blob_operations.py#L1702

It's sort of tough to tell if the code in `process_content` is actually sending requests but possibly if you try downloading a large payload in a single request (up to 32 MiB we expect only one request by default) and see if the pipeline gets hit multiple times., that may help to shed some light.

Comment (Stevenjin8) [reply sdk/storage/azure-storage-blob/azure/storage/blob/_download.py:214]: @jalauzon-msft I think I see what you mean, but I also can't seem to get streaming to run. I'll talk to @lmazuel about it tomorrow in our 1:1.

Comment (Stevenjin8) [reply sdk/storage/azure-storage-blob/azure/storage/blob/_download.py:214]: @jalauzon-msft I'm a bit lost and I think it might be better to leave things as they are until we figure out exactly what is going on with streaming (I can make an issue).

https://github.com/Azure/azure-sdk-for-python/blob/b0f9e02b2d597a57dcef33bb265c4b5c47ad0a85/sdk/storage/azure-storage-blob/azure/storage/blob/_download.py#L46

What I had in mind is to keep most of the code as it is and just replace the `requests` exceptions with `azure.core` exceptions. But if that line is not using the pipeline, then I would be tripling the retry mechanism. If that line is not using the pipeline, then I would be getting rid of retries all together.

**Review by lmazuel — APPROVED** [review]
LGTM

**Review by vincenttran-msft — APPROVED** [review]
LGTM! I see there is one hanging discussion -- if the resolution is to keep it as-is then I would say it is good to merge. If there is still a discussion to be had, Jacob is OOF and will be back after the long weekend + 1 (back online 7/6)

**Review by vincenttran-msft — COMMENTED** [review]
I spoke with Jacob briefly offline -- lets wait until we can have a further discussion next week before merging this in!

**Comment by Stevenjin8** [comment]
@vincenttran-msft @jalauzon-msft I think I figured out what is going on.

In the contentious line of code 

https://github.com/Azure/azure-sdk-for-python/blob/b0f9e02b2d597a57dcef33bb265c4b5c47ad0a85/sdk/storage/azure-storage-blob/azure/storage/blob/_download.py#L46

`data` is an instance of `azure.core.pipeline.transport._requests_basic.StreamDownloadGenerator`. Looking at the history of `StreamDownloadGenerator.__next__`, I found that

* until March 2021 the `__next__` method handled its own retries meaning that it caught `requests` exceptions
* #17078 removed the retry mechanism from the `__next__` method but did not catch any `requests` exceptions and reraise them as `azure.core` exceptions
* In response to #17078, PR #18164 added the current retry mechanism to catch `requests` exceptions thrown by `StreamDownloadGenerator.__next__`
* In theory, stream download generators should only throw `azure.core` exceptions. As such, #19433 and #20888 modified `StreamDownloadGenerator.__next__` to catch `requests` exceptions and raise them as `azure.core` exceptions. Crucially, these PRs didn't update the retry mechanism left behind by #18164, invalidating the retry mechanism.

Currently working on tests that confirm that this PR restores the behaviour back to #18164's intent. If there is consensus, open a PR to revert #25084.

Please let me know if you have any questions or concerns.

File: sdk/storage/azure-storage-blob/CHANGELOG.md
Comment (jalauzon-msft) [inline sdk/storage/azure-storage-blob/CHANGELOG.md:18]: Looks like maybe a bad merge here? You can remove all these.

File: sdk/storage/azure-storage-blob/azure/storage/blob/_download.py
Comment (jalauzon-msft) [reply sdk/storage/azure-storage-blob/azure/storage/blob/_download.py:216]: Given your investigation, I think this is the best course of action. It seems the response streaming no longer throws these `requests` errors and therefore, currently, our retry logic here is not doing anything. Additionally, looking into the implementation of `StreamDownloadGenerator` it seems to be reading the raw response, meaning it's likely not using our normal retry pipeline so I don't think we'll be tripling the retry count.

File: sdk/storage/azure-storage-blob/tests/test_get_blob.py
Comment (jalauzon-msft) [inline sdk/storage/azure-storage-blob/tests/test_get_blob.py:1023]: What's this test for? I don't think we need this?

If we do need it, you'll need to check in a recording.

File: sdk/storage/azure-storage-blob/tests/test_retry.py
Comment (jalauzon-msft) [inline sdk/storage/azure-storage-blob/tests/test_retry.py:441]: Does this need to be Live only?

**Review by jalauzon-msft — COMMENTED** [review]
Thanks @Stevenjin8 for investigating this further and providing your detailed summary! After seeing your investigation and doing a bit more myself, I think this is a good change.

Just left some minor review comments.

File: sdk/storage/azure-storage-blob/tests/test_get_blob.py
Comment (Stevenjin8) [reply sdk/storage/azure-storage-blob/tests/test_get_blob.py:1023]: oops

File: sdk/storage/azure-storage-blob/tests/test_retry.py
Comment (Stevenjin8) [reply sdk/storage/azure-storage-blob/tests/test_retry.py:441]: I think so? I'm really not sure but when I run it without the decorator, the test gets stuck. Maybe I could mock the transport? I'm open to suggestions.

Comment (jalauzon-msft) [reply sdk/storage/azure-storage-blob/tests/test_retry.py:441]: I don't see anything here that would cause a problem recording since you're mocking the response streaming. Did you try generating a recording?

- Remove the decorator, set `RUN_IN_LIVE` to `True`, and set `SKIP_LIVE_RECORDING` to `False` in the settings file.
- Run the test. It should generate a `.yaml` file.
- Then set `RUN_IN_LIVE` to `False` and run the test again.

**Review by jalauzon-msft — APPROVED** [review]
Thanks again for your work on this!
I've approved just pending the discussion on if the test can be recorded.

File: sdk/storage/azure-storage-blob/tests/test_retry.py
Comment (lmazuel) [reply sdk/storage/azure-storage-blob/tests/test_retry.py:441]: @mccoyp to the rescue if necessary

**Pull request merged**
